src/Cardano/Wallet/AddressDiscovery.hs

```diff
@@ -0,0 +1,229 @@
+defaultAddressPoolGap :: AddressPoolGap
+defaultAddressPoolGap =
+    AddressPoolGap 20
+
+
+-- ** Address Pool
+
+-- | An 'AddressPool' which keeps track of sequential addresses within a given
+-- Account and change chain. See 'mkAddressPool' to create a new or existing
+-- pool:
+--
+-- >>> mkAddressPool xpub gap changeChain mempty
+-- AddressPool { }
+data AddressPool = AddressPool
+    { accountPubKey
+        :: !(Key 'AccountK XPub)
+        -- ^ Corresponding key for the pool (a pool is tied to only one account)
+    , gap
+        :: !AddressPoolGap
+        -- ^ The actual gap for the pool. This can't change for a given pool.
+    , changeChain
+        :: !ChangeChain
+        -- ^ Whether this pool tracks addrs on the internal or external chain
+    , indexedAddresses
+        :: !(Map Address (Index 'Soft 'AddressK))
+    } deriving (Generic, Show, Eq)
+
+instance NFData AddressPool
+
+-- | Get all addresses in the pool, sorted from the first address discovered,
+-- up until the next one.
+--
+-- In practice, we always have:
+--
+-- > mkAddressPool key g cc (addresses pool) == pool
+addresses :: AddressPool -> [Address]
+addresses = map fst . sortOn snd . Map.toList . indexedAddresses
+
+-- | Create a new Address pool from a list of addresses. Note that, the list is
+-- expected to be ordered in sequence (first indexes, first in the list).
+--
+-- The pool will grow from the start if less than @g :: AddressPoolGap@ are
+-- given, such that, there are always @g@ undiscovered addresses in the pool.
+mkAddressPool
+    :: Key 'AccountK XPub
+    -> AddressPoolGap
+    -> ChangeChain
+    -> [Address]
+    -> AddressPool
+mkAddressPool key g cc addrs = AddressPool
+    { accountPubKey = key
+    , gap = g
+    , changeChain = cc
+    , indexedAddresses = nextAddresses key g cc minBound <>
+        Map.fromList (zip addrs [minBound..maxBound])
+    }
+
+
+-- | Lookup an address in the pool. When we find an address in a pool, the pool
+-- may be amended if the address was discovered near the edge. It is also
+-- possible that the pool is not amended at all - this happens in the case that
+-- an address is discovered 'far' from the edge.
+lookupAddress
+    :: Address
+    -> AddressPool
+    -> (Maybe (Index 'Soft 'AddressK), AddressPool)
+lookupAddress !target !pool =
+    case Map.lookup target (indexedAddresses pool) of
+        Just ix ->
+            (Just ix, extendAddressPool ix pool)
+        Nothing ->
+            (Nothing, pool)
+
+-- | If an address is discovered near the edge, we extend the address sequence,
+-- otherwise we return the pool untouched.
+extendAddressPool
+    :: Index 'Soft 'AddressK
+    -> AddressPool
+    -> AddressPool
+extendAddressPool !ix !pool
+    | isOnEdge  = pool { indexedAddresses = indexedAddresses pool <> next }
+    | otherwise = pool
+  where
+    edge = Map.size (indexedAddresses pool)
+    isOnEdge = edge - fromEnum ix <= fromEnum (gap pool)
```

Comment:
the definition of `isOnEdge` follows from the standard?

Reply:
Well, the specification is rather evasive about _how_ this should be handled. 

```
- scan addresses of the external chain; respect the gap limit described below
- if no transactions are found on the external chain, stop discovery
```

The approach above is how we've translated this, and works fairly well (cf property tests)

Reply:
:+1: